core/services/keeper/validate.go

```diff
@@ -1,34 +1,79 @@
 package keeper
 
 import (
+	"reflect"
+
 	"github.com/pelletier/go-toml"
 	"github.com/pkg/errors"
 	uuid "github.com/satori/go.uuid"
 
+	"github.com/smartcontractkit/chainlink/core/logger"
 	"github.com/smartcontractkit/chainlink/core/services/job"
+	"github.com/smartcontractkit/chainlink/core/services/pipeline"
 )
 
+// expectedObservationSourceRaw this is the expected observation source of the keeper job.
+const expectedObservationSourceRaw = `
+encode_check_upkeep_tx   [type=ethabiencode
+                          abi="checkUpkeep(uint256 id, address from)"
+                          data="{\"id\":$(jobSpec.upkeepID),\"from\":$(jobSpec.fromAddress)}"]
+check_upkeep_tx          [type=ethcall
+                          failEarly=true
+                          extractRevertReason=true
+                          contract="$(jobSpec.contractAddress)"
+                          gas="$(jobSpec.checkUpkeepGasLimit)"
+                          data="$(encode_check_upkeep_tx)"]
+decode_check_upkeep_tx   [type=ethabidecode
+                          abi="bytes memory performData, uint256 maxLinkPayment, uint256 gasLimit, uint256 adjustedGasWei, uint256 linkEth"]
+encode_perform_upkeep_tx [type=ethabiencode
+                          abi="performUpkeep(uint256 id, bytes calldata performData)"
+                          data="{\"id\": $(jobSpec.upkeepID),\"performData\":$(decode_check_upkeep_tx.performData)}"]
+perform_upkeep_tx        [type=ethtx
+                          minConfirmations=0
+                          to="$(jobSpec.contractAddress)"
+                          data="$(encode_perform_upkeep_tx)"
+                          gasLimit="$(jobSpec.performUpkeepGasLimit)"
+                          txMeta="{\"jobID\":$(jobSpec.jobID)}"]
+encode_check_upkeep_tx -> check_upkeep_tx -> decode_check_upkeep_tx -> encode_perform_upkeep_tx -> perform_upkeep_tx`
+
+// expectedPipeline it is basically parsed expectedObservationSourceRaw value
+var expectedPipeline pipeline.Pipeline
+
+func init() {
+	pp, err := pipeline.Parse(expectedObservationSourceRaw)
+	if err != nil {
```

Comment:
Fatal log stops an app but in a more accurate way.

Reply:
you're right! logger.Fatal looks good.

Reply:
`Logger` is not guaranteed to be set up correctly here. I'd prefer to panic to be honest.

Reply:
@samsondav according to this [logic](https://github.com/smartcontractkit/chainlink/blob/develop/core/logger/default.go#L18-L57) we make sure the `Default` logger is setup correctly, otherwise, the app will never be running. Moreover, we use this logger in other parts of the app. `Fatal` function stops the app with an understandable message. I've tested it on my local env. Could you share some details how to reproduce a case when this `Fatal` execution works not correctly?